### PR TITLE
feat: LWW-Register impl & Tests

### LINKED ISSUE #81: Implement built-in CRDTs: LWW Register
### Background
Register types are useful for CRO to specify a single value that can be updated concurrently by multiple nodes. LWW (Last-Write-Win) Register is a basic type for this purpose.

### Task
- implement
- test

### Resources
- https://lars.hupel.info/topics/crdt/07-deletion/
- https://jsonjoy.com/specs/json-crdt/model-document/crdt-algorithms see the section on "Last-Write-Wins (LWW) CRDT Algorithm"
- https://www.bartoszsypytkowski.com/operation-based-crdts-registers-and-sets/

### PR DESCRIPTION
**Pull request by joaopereira12**
resolves #81

### CODE REVIEW

File: packages/crdt/src/builtins/LWWRegister/index.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+type Timestamp = [timestamp: number, replicaId: string];
```
Comment (d-roak) [inline packages/crdt/src/builtins/LWWRegister/index.ts:1]: don't use `replicaId`, use `nodeId`

File: packages/crdt/src/builtins/LWWRegister/index.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    
+
+    constructor(element: T, replicaId: string) {
+        this._element = element;
+        this._timestamp = [Date.now(), replicaId];
+    }
+
+    assign(element: T, replicaId: string): void {
+        const timestamp = Date.now();
+        if(compareTimestamps(this._timestamp, [timestamp, replicaId])) {
```
Comment (d-roak) [inline packages/crdt/src/builtins/LWWRegister/index.ts:15]: the assign is supposed to run locally, the timestamp will never be smaller

File: packages/crdt/src/builtins/LWWRegister/index.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    getElement(): T {
+        return this._element;
+    }
+
+    getTimestamp(): Timestamp {
+        return this._timestamp;
+    }
+
+    compare(register: LWWRegister<T>): boolean {
+        return (this._timestamp <= register.getTimestamp());
```
Comment (d-roak) [inline packages/crdt/src/builtins/LWWRegister/index.ts:30]: this is wrong

File: packages/crdt/src/builtins/LWWRegister/index.ts (hunk 1 above)
Comment (d-roak) [reply packages/crdt/src/builtins/LWWRegister/index.ts:1]: don't create this type. Store both timestamp and nodeId in 2 vars within the class

File: packages/crdt/src/builtins/LWWRegister/index.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    merge(register: LWWRegister<T>): void {
+        if(compareTimestamps(this._timestamp, register.getTimestamp())) {
```
Comment (d-roak) [inline packages/crdt/src/builtins/LWWRegister/index.ts:34]: extend the if logic here and delete the auxiliary function

**Review by d-roak — CHANGES_REQUESTED** [review]

File: packages/crdt/src/builtins/LWWRegister/index.ts (hunk 3 above)
Comment (joaopereira12) [reply packages/crdt/src/builtins/LWWRegister/index.ts:30]: in this paper: https://dsf.berkeley.edu/cs286/papers/crdt-tr2011.pdf - they implement the compare function this way.
But i've seen some other solutions in the resources on the issue that don't have this function.

Comment (d-roak) [reply packages/crdt/src/builtins/LWWRegister/index.ts:31]: timestamp is not a number, it's a tuple. that's why this is wrong

File: packages/crdt/tests/LWWRegister.test.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        expect(register1.getNodeId()).toEqual("node2");
+    });
+
+    // test('Test Compare', () => {
+    //     let register1 = new LWWRegister<string>("alice", "node1");
+    //     let register2 = new LWWRegister<string>("alice", "node2");
+
+    //     expect(register1.compare(register2)).toEqual(true);
+    //     expect(register2.compare(register1)).toEqual(false);
+    // });
```
Comment (d-roak) [inline packages/crdt/tests/LWWRegister.test.ts:21]: fix this

File: packages/crdt/src/builtins/LWWRegister/index.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    merge(register: LWWRegister<T>): void {
+        const otherTimestamp = register.getTimestamp();
+        const otherNodeId = register.getNodeId();
+        if (otherTimestamp > this._timestamp ||
+            (otherTimestamp === this._timestamp && otherNodeId > this._nodeId)) {
+            this._element = register.getElement();
+            this._timestamp = otherTimestamp;
+            this._nodeId = otherNodeId;
+        }
```
Comment (d-roak) [inline packages/crdt/src/builtins/LWWRegister/index.ts:42]: change to return prematurely if:
- the timestamp is bigger
- timestamp equal and nodeId lower

**Review by d-roak — CHANGES_REQUESTED** [review]

**Review by d-roak — APPROVED** [review]
lgtm

**Pull request merged**